include/tilck/mods/irqchip.h

```diff
@@ -0,0 +1,101 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+
+#pragma once
+
+#include <tilck/common/basic_defs.h>
+#include <tilck/kernel/list.h>
+
+enum irq_type {
+   IRQ_TYPE_NONE           = 0x00000000,
+   IRQ_TYPE_EDGE_RISING    = 0x00000001,
+   IRQ_TYPE_EDGE_FALLING   = 0x00000002,
+   IRQ_TYPE_EDGE_BOTH      = (IRQ_TYPE_EDGE_FALLING | IRQ_TYPE_EDGE_RISING),
+   IRQ_TYPE_LEVEL_HIGH     = 0x00000004,
+   IRQ_TYPE_LEVEL_LOW      = 0x00000008,
+};
+
+struct fdt_irq_param {
+   int intc_node;
+   int param_nums;
+#define FDT_MAX_INTERRUPTS_PARAMS 16
+   u32 params[FDT_MAX_INTERRUPTS_PARAMS];
+};
+
+struct fdt_irqchip {
+   const struct fdt_match *id_table;
+   int (*init)(void *fdt, int node, const struct fdt_match *match);
+};
+
+struct fdt_irqchip_ops {
+   void (*hwirq_set_mask)(int hwirq, void *priv);
+   void (*hwirq_clear_mask)(int hwirq, void *priv);
+   bool (*hwirq_is_masked)(int hwirq, void *priv);
+   void (*hwirq_set_type)(int hwirq, int type, void *priv);
+   int (*xlate_irq_param)(struct fdt_irq_param *param,
+                          int *hwirq, int *type, void *priv);
+};
+
+struct irq_domain {
+   void *priv;
+   struct list_node node;
+   int fdt_node;
+   struct fdt_irqchip_ops *ops;
+   size_t irq_map_size;
+   int irq_map[];
+};
+
+struct irq_data {
+   bool present;
+   u32 hwirq;
+   enum irq_type type;
+   u32 unhandled_count;
+   struct irq_domain *domain;
+};
+
+#define DEFINE_FDT_IRQCHIP(_name)                             \
+                     struct fdt_irqchip fdt_irqchip_##_name   \
+                                      __attribute__((used))   \
+        __attribute__((section (".modules.fdt_irqchip.2")))
+
+#define FDT_IRQCHIP_START                                     \
+({                                                            \
+                         static struct fdt_irqchip start[0]   \
+                          __attribute__((aligned(NBITS/8)))   \
+                                      __attribute__((used))   \
+        __attribute__((section (".modules.fdt_irqchip.1")));  \
```

Comment:
First of all, explain`.simplebus_driver`, this is supposed to be useless code that I forgot to clean up, and I will delete it immediately.
For`.modules.*`, I have to make a clarification here. First, these segments have nothing to do with the bootloader or fdt. You can observe the `struct fdt_irqchip`, which actually represents the **driver** of an interrupt controller (which records the address of the driver initialization function and the name of the matching device), rather than an abstraction of the interrupt controller **device**;
The interrupt controllers **driver** register themselves with `DEFINE_FDT_IRQCHIP`, and then we combine these `struct fdt_irqchip` into an array of `.modules.fdt_irqchi.2` via the link script; In this way, `init_fdt_irqchip()` can look up the interrupt controller nodes in the device tree in turn, and then iterate through the `.modules.fdt_irqchi.2` array for each node to see if it can find a matching **driver**, and call the driver initialization function if there is one. So `init_fdt_irqchip()` needs to use `FDT_IRQCHIP_START` to know where the `.modules.fdt_irqchi.2` array starts and ends, This cannot be done using tilck's existing modules initialization hooks alone.

Reply:
@aenrbes If that has nothing to do with the bootloader, then it's much easier. You can mark as many functions as you need with `__attribute__((constructor))`: they will be automatically called during the early initialization in kmain(). In those functions, each driver can append a struct of its own on a global linked list. You really don't need any sort of fancy extra section for things like that. Look at the logic for selftests:

```
struct self_test {
   struct list_node node;
   const char *name;
   enum se_kind kind;
   void (*func)(void);
};

#define REGISTER_SELF_TEST(__name, __kind, __func)  \
                                                                \
   static struct self_test se_##__name##_inst = {               \
      .name = #__name,                                          \
      .kind = __kind,                                           \
      .func = __func                                            \
   };                                                           \
                                                                \
   __attribute__((constructor))                                 \
   static void __register_se_##__name(void) {                   \
      se_register(&se_##__name##_inst);                         \
   }
```

`se_register()` is just:
```
void se_register(struct self_test *se)
{
   list_add_tail(&se_list, &se->node);
}
```


The REGISTER_SELF_TEST() macro used like this:
```
REGISTER_SELF_TEST(kmalloc_perf, se_long, &selftest_kmalloc_perf)
```

**Note:** you probably don't even need all the macro stuff here. You just need to a constructor function that simply appends an "ops" struct to a linked list.

Once you have a linked list with structs, you can probe all the drivers, one by one. It's flexible because it doesn't require a single place where all the drivers are declared upfront and it doesn't use any other custom section other than the `ctors` one, that is a well-known section used by compilers.

Isn't your problem **completely solvable** with this approach? If not, why?

Reply:
Thank you for pointing it out and I will change it according to this approach.

Reply:
Thank you for understanding and working on a fix.